corehq/apps/locations/adjacencylist.py

```diff
@@ -271,3 +271,14 @@ def get_descendants(self, **kw):
         with timing("cte"):
             cte_set = type(self).objects.cte_get_descendants(self, **kw)
         return ComparedQuerySet(mptt_set, cte_set, timing)
+
+
+def _is_empty(queryset):
+    query = queryset.query
+    if query.is_empty():
+        return True
+    try:
+        query.sql_with_params()
+    except EmptyResultSet:
```

Comment:
This seems like a non-sequitor, how/why would `sql_with_params` raise this exception?  Does it do some analysis of the generated sql to determine that it'll return no results so it doesn't need to run it?

Reply:
Yes. It's unfortunate that it doesn't do this check directly in `query.is_empty()`. When it generates the SQL it analyzes the `WHERE` clause elements and if, for example, there is an `IN` clause with an empty list it will raise this exception. In the case of a recursive CTE, which uses a `UNION`, the non-recursive clause of the union would be dropped because Django is trying to optimize the query, but in this case the optimization results in an invalid query.